### PR TITLE
Refactored sampled softmax

### PR DESCRIPTION
**Pull request by eschnou**
The current sampled softmax implementation doesn't improve speed. This can be linked to this issue in Tensorflow seq2seq sample: https://github.com/tensorflow/tensorflow/issues/4138 which was fixed by this commit: https://github.com/tensorflow/tensorflow/pull/4270/commits/e5c81310ec2345cb9bd8f5c8ebb5e90724e0458c

I've reproduced the fix on our codebase and could observe a major speedup when using sampled softmax:

Hiddensize 1024 
- no sampled softmax      1.1s/it
- sampledSoftmax=128 - 1.9s/it

### CODE REVIEW

**Comment by eschnou** [comment]
Don't merge yet. I've now trained a full model and have weird results. Not sure of the cause yet, but it might be this last change I made. Will let you know.

**Comment by eschnou** [comment]
It seems the training is smooth (loss decreasing at similar speed that previously), but the interactive test gives really poor results mostly empty strings. So, I assume there is an issue on the testing side, but I can't find it. Will continue to investigate.

**Comment by Conchylicultor** [comment]
I'm not sure your last commit is a good idea. The sampling softmax is only used by the loss function. On testing, the projection is done here:
https://github.com/eschnou/DeepQA/blob/98f0af28a4ed42bd52fa74e9cc5f1f3db1f19670/chatbot/model.py#L190

If you disable `outputProjection` for testing, I'm not sure how your model now how to reuse the weights that you trained. Because this line will be deactivated, the model probably initialize random weights at testing time.
https://github.com/eschnou/DeepQA/blob/98f0af28a4ed42bd52fa74e9cc5f1f3db1f19670/chatbot/model.py#L178

**Comment by eschnou** [comment]
Yes you are correct, I've reverted it. This was just a test on my local branch and I had not realised it was also pushed on the pull request. So:

- on one hand I think the fix really works and improves learning speed, after all, I've just re-integrated the same change that Google did.

- on the other hand I have weird results when testing, but this might come from another cause. I need to redo a test with controlled params, comparing results from master brancnch vs this patch.

**Comment by Conchylicultor** [comment]
Your changes seems correct so I'll say your second point is more likely. Maybe the problem comes from somewhere else in DeepQA.

**Comment by eschnou** [comment]
I've just done a test using autoencode, comparing results of:

- python3 main.py --autoEncode
- python3 main.py --autoEncode --softmaxSamples=1024

Softmax speedup is x2
Results are similar when testing (autoencode ok in both cases)

-> I think my problem was somewhere else and I suggest you merge this one if you are ok with it.

**Comment by Conchylicultor** [comment]
Great. It's done.
